lib/hash/utils.js

```diff
@@ -13,14 +13,32 @@ function toArray(msg, enc) {
   var res = [];
   if (typeof msg === 'string') {
     if (!enc) {
+      // Inspired by stringToUtf8ByteArray() in closure-library by Google
+      // https://github.com/google/closure-library/blob/8598d87242af59aac233270742c8984e2b2bdbe0/closure/goog/crypt/crypt.js#L117-L143
+      // Apache License 2.0
+      // https://github.com/google/closure-library/blob/master/LICENSE
+      let p = 0;
       for (var i = 0; i < msg.length; i++) {
         var c = msg.charCodeAt(i);
-        var hi = c >> 8;
-        var lo = c & 0xff;
-        if (hi)
-          res.push(hi, lo);
-        else
-          res.push(lo);
+        if (c < 128) {
+          res[p++] = c;
+        } else if (c < 2048) {
+          res[p++] = (c >> 6) | 192;
+          res[p++] = (c & 63) | 128;
+        } else if (
```

Comment:
What do you think about having a function `isSurrogatePair()`?

Reply:
Where else can we use it? IMO if the only application of `isSurrogatePair()` remains within one `if`, extracting a function won't add much value. However, it will definitely carry a bit of a performance overhead, which a vote against doing so.

The whole code block that generates the utf8 byte array from a JS utf16 string is pretty low-level. Making one tiny bit in it a bit more clear won't make the rest of the lines less cryptic 😅

Reply:
It won't carry any performance overhead :wink: V8 (and other JS engines) are clever enough to inline it. What's going to better, though, is that the code at this line will become much more readable.

Reply:
```js
if ((c & 0xFC00) == 0xD800) && (i + 1) < msg.length &&
            ((msg.charCodeAt(i + 1) & 0xFC00) == 0xDC00)) {
  // ...
}
```

↓

```js
const isSurrogatePair = (msg, i) =>
  (msg.charCodeAt(i) & 0xFC00) == 0xD800)
  && (i + 1) < msg.length && ((msg.charCodeAt(i + 1) & 0xFC00) == 0xDC00)

// ...

if (isSurrogatePair(msg, i)) {
  // ...
}
```

? 🤔

Reply:
How about:
```js
function isSurrogatePair(msg, i) {
  if ((msg.charCodeAt(i) & 0xFC00) !== 0xD800) {
    return false;
  }
  if (i + 1 >= msg.length) {
    return false;
  }
  return (msg.charCodeAt(i + 1) & 0xFC00) === 0xDC00;
}
```

This function could be defined in a top-level scope in this file.

Reply:
👌 

Will just change this:

```diff
 if (i + 1 >= msg.length) {
 if (i < 0 || i + 1 >= msg.length) {
   return false;
 }
```